### ISSUE TITLE
File or attachment user input action in Composer v2.1.1 not working

### ISSUE THREAD

**Issue by anujprasad** [microsoft/BotFramework-Composer#8867]
I am trying to create an attachment input action in bot framework composer and followed [these steps
](https://docs.microsoft.com/en-us/composer/how-to-ask-for-user-input?tabs=v2x#create-an-attachment-input-action) but I am getting  the error message - "Object reference not set to an instance of an object" on emulator vesrion 4.14.0. Please find below the screen shot of the implementation and error message:

![image](https://user-images.githubusercontent.com/38533303/146875344-a79957ea-1fdd-49c2-8d10-b47a1f49f37b.png)

![image](https://user-images.githubusercontent.com/38533303/146875614-1ab10f93-1ff9-4276-bcf1-6adcbe43fdc4.png)

**Comment by anujprasad** [comment]
My Bot framework Composer version details:

Version:  2.1.1
Electron: 8.2.4
Chrome: 80.0.3987.165
NodeJS: 12.13.0
V8: 8.0.426.27-electron.0

**Comment by ramfattah** [comment]
Hey @anujprasad, is your intent to attach a file? It looks like your getting the error as the input from the card is a text rather than an attachment file. Run your bot and click on the upload file (paper clip icon) button on bottom left of the chat to attach a file. Please let me know if this helps.

**Comment by anujprasad** [comment]
@ramfattah  Yes, my intent is to attach a pdf file. The input text "number" that you see in the screen shot is just a regular expression trigger to trigger the number dialog and the very first step on that dialog is to ask user for file attachment. It throws the error "Object reference not set to an instance of an object" even before prompting the user the message "Please upload your invoice". Ideally user should first see this prompt "Please upload your invoice" and then he/she would click on upload file (paper clip icon) button to attach a file, it should not throw the exception before that. 

I have shared the composer authoring screen for "Prompt for a file and attachment" , do you find any anomaly in that implementation that could throw the exception - "Object reference not set to an instance of an object"?

**Comment by anujprasad** [comment]
Re-attaching the composer bot authoring screen:

![image](https://user-images.githubusercontent.com/38533303/147039308-c10fc361-ca62-4ab2-b2b5-74c235f191e6.png)

**Comment by drmanhatin** [comment]
I'm facing the same issue.

**Comment by drmanhatin** [comment]
In my case the input prompt did not have any text, this is why it errored. Ui should reflect it being necessary or should be possible to not contain any value.

**Comment by anujprasad** [comment]
I figured out the root cause and solution, incase anyone facing similar issue:
Exact steps to reproduce the issue:

1. Select the + icon then select File or attachment from the Ask a question menu
2. Enter Please send an image. in the Prompt field of the Properties pane.
3. Switch to the User Input tab ()
- Enter dialog.attachments in the Property field.
- Enter all in the Output format field.
- Enter "application/pdf" in the contentType. **_Note that contentType field is marked as mandatory_**
4. Now run your bot and invoke the dialog that has prompt for file or an attachment - your get the error message - "Object reference not set to an instance of an object"

The root cause for the issue was contentType which is marked as mandatory field with red *, now even if I remove the text "application/pdf" from contentType field I continue to get the error message - "Object reference not set to an instance of an object". So the fix is to remove the control "prompt for file or an attachment" first and then re-add it and make sure you have not added any text in the mandatory field - "contentType " :). 


![image](https://user-images.githubusercontent.com/38533303/147531847-9113fa88-07fb-4c0b-b845-0ec86cec4510.png)

**Comment by anujprasad** [comment]
Quick question - Once the file is uploaded by user using the prompt "File or attachment input", does that file stays in the server forever or it's stored in memory within dialog scope? If I have to hand it off to any live agent such as customer helpdesk, what's the best possible way to store the transcript and this file and then send it over an email to the live agent?

**Comment by ramfattah** [comment]
Hi @anujprasad , Microsoft Bot Framework is an open platform that allows developers to integrate with a variety of agent engagement platforms. This is the official [Microsoft Documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-design-pattern-handoff-human?view=azure-bot-service-4.0#:~:text=Transition%20conversations%20from%20bot%20to%20human) on Handoff integration models to provide a common language and best practices for bot developers and system integrators building conversational AI systems with human in the loop.

Also, Here is the [Bot Framework v4 Handoff Library](https://github.com/microsoft/BotBuilder-Samples/tree/main/experimental/handoff-library).

**Comment by ParthHadkar1203** [comment]
> Hi @anujprasad , Microsoft Bot Framework is an open platform that allows developers to integrate with a variety of agent engagement platforms. This is the official [Microsoft Documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-design-pattern-handoff-human?view=azure-bot-service-4.0#:~:text=Transition%20conversations%20from%20bot%20to%20human) on Handoff integration models to provide a common language and best practices for bot developers and system integrators building conversational AI systems with human in the loop.
> 
> Also, Here is the [Bot Framework v4 Handoff Library](https://github.com/microsoft/BotBuilder-Samples/tree/main/experimental/handoff-library).

can i get upload file url so i can store some where in api and the through api i will dowload it

**Comment by ParthHadkar1203** [comment]
if user send me pdf attachment in bot composer.
How will I get to know file is attach or not?
so i can store it some where and then download from it for example (azure blob storage).

**Comment by ParthHadkar1203** [comment]
> if user send me pdf attachment in bot composer. How I will store that pdf in my code. How it will convert to base 64 or in URL to send via API to store somewhere

My issue is resolved and it uploading in my blob

**Comment by ParthHadkar1203** [comment]
how to used validation for file ie. file should be in .pdf format only in bot composer

**Comment by RaviKasaudhan** [comment]
> > if user send me pdf attachment in bot composer. How I will store that pdf in my code. How it will convert to base 64 or in URL to send via API to store somewhere
> 
> My issue is resolved and it uploading in my blob storage and getting the URL too. NOTE : Storing locally only
> 
> When I try to run through server its show error Below Error : An attempt was made to access a socket in a way forbidden by its access permissions. An attempt was made to access a socket in a way forbidden by its access permissions.

Hi Hadkar, Can you please explain the process you followed so that whatever the user uploads, you get the URL in blob storage? Right now, after the user uploads any pdf i get ContentURL which is having localhost link. After copy-pasting that localhost link to any website its downloading the file without the extension. Can you also say how did you solve this? 
Any help is much appreciated.

**Comment by ParthHadkar1203** [comment]
See in my case I get the whole name i.e - filename.pdf etc
In your case exactly what link are you getting?
and my issue is still pending

**Comment by chinubansal2014** [comment]
How to store this file to azure blob storage or sharepoint list ?

**Comment by surlicki** [comment]
I'm also searching the possibility of saving attachments to azure storage. My bot as channel is using Teams. What I done up to now - I used "prompt for a file or attachment" and get content url of the file. Im looking for the next steps. From web I found that file should be sent via http request to azure blob. Is it a proper way, should I use this content url? Is it possible to done form Bot Framework Composer or extra coding is necessary? I appreciate all advices.
